### PR TITLE
Bulk scan/fix for insecure content from the admin

### LINKED ISSUE #58: Add ability to run a bulk scan/fix from the admin
**Is your enhancement related to a problem? Please describe.**
In #39 a WP-CLI script was added to scan for insecure items and attempt to fix those. This is a nice approach to bulk fix content for sites that may have been converted from insecure to secure but still have references to insecure items.

Not all users will have the access or capability to run a WP-CLI command though, so would be great to add an option in the admin to run the same functionality.

**Describe the solution you'd like**
Have a place somewhere in the admin where you can initiate a scan for insecure items. Might want some options like the content type to scan and the number of items to scan at a time and maybe an autofix option. This would then run the scan and let you know the items that it finds (and if autofix is on, if the items were able to be fixed).

I can see a similar approach to what is done in [ElasticPress](https://github.com/10up/ElasticPress), if we want to use that as a reference.

### PR DESCRIPTION
**Pull request by kmgalanakis**
### Description of the Change

Closes #58

### How to test the Change

* Create a few posts with the following WP cli command `echo '<!-- wp:image {"sizeSlug":"large"} --><figure class="wp-block-image size-large"></figure><!-- /wp:image --><!-- wp:image {"sizeSlug":"large"} --><figure class="wp-block-image size-large"></figure><!-- /wp:image --><!-- wp:image {"sizeSlug":"large"} --><figure class="wp-block-image size-large"></figure><!-- /wp:image -->' | wp post generate --count=1 --post_type=car`
* The newly created posts will contain insecure content (image). In the Admin, go to `Tools -> Insecure Content Warning` and run (and dry-run) the process to fix the newly created posts. Play with all the options.
* Verify the the already existing WP cli command to fix insecure content in posts still works.

### Changelog Entry

> Added - New admin screen to fix insecure content in posts.

### Checklist:

- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests pass.

### CODE REVIEW

**Comment by peterwilsoncc** [comment]
@kmgalanakis Feel free to ping me on the ticket once this is ready for review so it doesn't get lost.

I'm happy to do an early scan if you wish, too, just say the word.

**Comment by kmgalanakis** [comment]
> @kmgalanakis Feel free to ping me on the ticket once this is ready for review so it doesn't get lost.
> 
> I'm happy to do an early scan if you wish, too, just say the word.

@peterwilsoncc the biggest part of the work here is done but the problem is that I realized my current approach is not performing very well when there is a large number of posts with insecure content to be updated. I want to change that and as soon as it's ready I will let you know.

**Comment by kmgalanakis** [comment]
> @kmgalanakis Feel free to ping me on the ticket once this is ready for review so it doesn't get lost.
> 
> I'm happy to do an early scan if you wish, too, just say the word.

@peterwilsoncc the fix is ready for review. I can still see the E2E tests failing but they are also failing on `develop` and I don't know how to fix them. I would be more than happy to also add the E2E test for this new feature but I don't know when I will be able to focus on this again, so I'm ok to do it after this feature is merged.

Please let me know what you think about the approach and the code.

Thank you.

File: includes/admin.php
Comment (peterwilsoncc) [inline includes/admin.php]:
```suggestion
		'edit_posts',
```

As a general capability, this seems closes to what the page is doing.

File: includes/classes/class-fixinsecurecontent.php
Comment (peterwilsoncc) [inline includes/classes/class-fixinsecurecontent.php]: I don't understand why the wait, what is the intent?

File: includes/classes/class-fixinsecurecontent.php
Comment (peterwilsoncc) [inline includes/classes/class-fixinsecurecontent.php]:
```suggestion
				$output         .= '<span class="warning">' . __( 'Warning', 'insecure-content-warning' ) . '</span>: ' . wp_strip_all_tags( $warning_message ) . PHP_EOL;
```

I know it isn't needed now but it protects ourselves from our future selves.

File: includes/classes/class-fixinsecurecontent.php
Comment (peterwilsoncc) [inline includes/classes/class-fixinsecurecontent.php]: Translator notes are intended to describe the placeholder rather than the string.

```suggestion
			// translators: %d The Post ID being scanned for insecure content.
			$message = sprintf( __( 'Unable to fetch details for post %d', 'insecure-content-warning' ), $post_id );
```

File: includes/classes/class-fixinsecurecontent.php
Comment (peterwilsoncc) [inline includes/classes/class-fixinsecurecontent.php]:
```suggestion
```

File: includes/classes/class-fixinsecurecontent.php
Comment (peterwilsoncc) [inline includes/classes/class-fixinsecurecontent.php]: As above re translator note and use of `%d`

File: includes/classes/class-fixinsecurecontent.php
Comment (peterwilsoncc) [inline includes/classes/class-fixinsecurecontent.php]: As above re translator note and use of `%d`

File: includes/classes/class-fixinsecurecontent.php
Comment (peterwilsoncc) [inline includes/classes/class-fixinsecurecontent.php]: As above re translator note and use of `%d`

File: includes/classes/class-fixinsecurecontent.php
Comment (peterwilsoncc) [inline includes/classes/class-fixinsecurecontent.php:220]: For non WP_CLI requests, this will need a capability check

```suggestion
		$current_post = get_post( $post_id );
		
		if ( ! defined( 'WP_CLI' ) || ! WP_CLI ) {
			if ( ! current_user_can( 'edit_post', $post_id ) ) {
				// Return early, show error user can't edit the post.
			}
		}
```

File: includes/classes/class-fixinsecurecontent.php
Comment (peterwilsoncc) [inline includes/classes/class-fixinsecurecontent.php]:
```suggestion
			// translators: 1: Number of insecure URLs updated, 2: Number of insecure URLs in post, 3: The post ID.
```

File: includes/classes/class-fixinsecurecontent.php
Comment (peterwilsoncc) [inline includes/classes/class-fixinsecurecontent.php]: As above

File: includes/partials/admin-page.php
Comment (peterwilsoncc) [inline includes/partials/admin-page.php]: No need to escape the constants, just variables and translations/

```suggestion
						<option selected="selected" value="all"><?php esc_html_e( 'All', 'insecure-content-warning' ); ?></option>
						<option value="posts"><?php esc_html_e( 'Individual post(s)', 'insecure-content-warning' ); ?></option>
						<option value="all_from_post_type"><?php esc_html_e( 'All from post type', 'insecure-content-warning' ); ?></option>
```

File: includes/rest.php
Comment (peterwilsoncc) [inline includes/rest.php]:
```suggestion
			'permission_callback' => 'edit_posts',
```

This might end up needing to be a little more complex to allow for CPTs registered with custom capabilities.

As it's editing the posts, it will definitely need to include a capability check and should probably include a nonce check for users without an application password.

File: includes/rest.php
Comment (peterwilsoncc) [inline includes/rest.php]:
```suggestion
			'permission_callback' => 'edit_posts',
```

Reveals data.

File: includes/rest.php
Comment (peterwilsoncc) [inline includes/rest.php:221]: I suggest using the REST APIs schema for this.

**Review by peterwilsoncc — CHANGES_REQUESTED** [review]
I've added some notes inline, there's a couple of security pick ups in there that are blockers.

There's a lot of unrelated changes in this PR due to the eslint fixes, while that should probably be done, I'd rather do it on a separate ticket to avoid confiscation. Especially given the introduction of bulk edits/rest endpoints that need a good security review.

File: includes/admin.php
Comment (kmgalanakis) [reply includes/admin.php]: Good point, thank you!

File: includes/classes/class-fixinsecurecontent.php
Comment (kmgalanakis) [reply includes/classes/class-fixinsecurecontent.php]: I can't remember exactly because it's been a while but I'm guessing this was a leftover from my initial approach. I removed it.

File: includes/classes/class-fixinsecurecontent.php
Comment (kmgalanakis) [reply includes/classes/class-fixinsecurecontent.php]: Better safe than sorry, thank you.

File: includes/classes/class-fixinsecurecontent.php
Comment (kmgalanakis) [reply includes/classes/class-fixinsecurecontent.php:220]: Good point, thank you.

File: includes/classes/class-fixinsecurecontent.php
Comment (kmgalanakis) [reply includes/classes/class-fixinsecurecontent.php]: Done, thank you.

File: includes/classes/class-fixinsecurecontent.php
Comment (kmgalanakis) [reply includes/classes/class-fixinsecurecontent.php]: Done, thank you.

File: includes/classes/class-fixinsecurecontent.php
Comment (kmgalanakis) [reply includes/classes/class-fixinsecurecontent.php]: Done, thank you.

File: includes/classes/class-fixinsecurecontent.php
Comment (kmgalanakis) [reply includes/classes/class-fixinsecurecontent.php]: Done, thank you.

File: includes/classes/class-fixinsecurecontent.php
Comment (kmgalanakis) [reply includes/classes/class-fixinsecurecontent.php]: Done, thank you.

File: includes/classes/class-fixinsecurecontent.php
Comment (kmgalanakis) [reply includes/classes/class-fixinsecurecontent.php]: Done, thank you.

File: includes/classes/class-fixinsecurecontent.php
Comment (kmgalanakis) [reply includes/classes/class-fixinsecurecontent.php]: Done, thank you.

File: includes/partials/admin-page.php
Comment (kmgalanakis) [reply includes/partials/admin-page.php]: Good point, thank you.

File: includes/rest.php
Comment (kmgalanakis) [reply includes/rest.php]: Good point, thank you.

File: includes/rest.php
Comment (kmgalanakis) [reply includes/rest.php]: Should I take the time to research this now?

File: includes/rest.php
Comment (kmgalanakis) [reply includes/rest.php:221]: This method does nothing more than prepare the params coming from the request for the methods that implement that counting and the fixing of posts with insecure content. How do you think we could use the REST APIs schema for this?

Comment (peterwilsoncc) [reply includes/rest.php:221]: Generally, the goal is to do it the WordPress way -- using schema gets a lot of things for free.

Setting the parameters in the schema will use the validation and sanitization built in to WP, so specifying the expected types there will automatically throw warnings and validation errors for invalid values. 

If WP discovers a security error in the RES API, by using the schema we can ensure that the upstream fixes are applied to the endpoints here too.

For post types, you'll probably need to do a little more validation -- see this line [onwards in distributor ](https://github.com/10up/distributor/blob/876f4b66e8e2db676748890139fce0ad5c4a1093/includes/rest-api.php#L210) -- to make sure the post type is intended to be visible in the rest api, the user has access, etc.

File: src/js/classic-editor.js
Comment (peterwilsoncc) [inline src/js/classic-editor.js:5]: At the start of the file you'll need to add:

```js
// eslint-disable-next-line import/no-unresolved
import { jQuery } from 'jquery';
```

This will ensure the dependency is included in `dist/js/classic-editor.asset.php` by [`@wordpress/dependency-extraction-webpack-plugin`](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-dependency-extraction-webpack-plugin/)

File: src/js/gutenberg.js
Comment (peterwilsoncc) [inline src/js/gutenberg.js:10]: As above.

File: src/js/utils/blur-insecure.js
Comment (peterwilsoncc) [inline src/js/utils/blur-insecure.js:1]: As above.

File: src/js/utils/check-content.js
Comment (peterwilsoncc) [inline src/js/utils/check-content.js:5]: As above.

File: src/js/utils/find-elements.js
Comment (peterwilsoncc) [inline src/js/utils/find-elements.js:3]: As above.

File: src/js/utils/replace.js
Comment (peterwilsoncc) [inline src/js/utils/replace.js:1]: As above.

File: includes/classes/class-fixinsecurecontent.php
Comment (peterwilsoncc) [inline includes/classes/class-fixinsecurecontent.php]: This should be limited to only posts available via the rest api -- per note re:schema. Many e-commerce plugins use post types for orders sow we definitely need to exclude those.

File: includes/classes/class-fixinsecurecontent.php
Comment (peterwilsoncc) [inline includes/classes/class-fixinsecurecontent.php]: **maybe**, _maybe_, maybe???

[Per MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/HEAD), this should tell you whether the file exists without having to download the full content. It should return the HTTP headers as-though you were making a GET request.

```suggestion
		$response              = wp_remote_head( $ssl );
```

File: includes/rest.php
Comment (kmgalanakis) [reply includes/rest.php:221]: The functionality to bulk scan/fix insecure content from the admin is actually a replication of the WP CLI functionality. The WP CLI functionality can properly scan/fix post types that are not shown in REST, so in my opinion, the admin functionality should behave exactly the same. According to my tests, neither the WP CLI nor the built admin functionality has problems scanning/fixing posts from a post type that is not shown in REST.

What I don't understand from your comment is how this specific function here could be avoided by the use of the WP REST API. Let me explain once again here that what this method does is "transform" the parameters of the WP REST API endpoint call into suitable parameters for the methods that count/scan/fix posts with insecure content. You see, these methods are also used from the WP CLI command and it wouldn't make too much sense to have a 1:1 relation between the REST API endpoints parameters with the parameters fo those methods, that is why I'm using this "transformation" method.

Still, I took the time to implement some more validation and sanitization callbacks for the actual REST API endpoint parameters to play more on the safe side.

File: includes/classes/class-fixinsecurecontent.php
Comment (kmgalanakis) [reply includes/classes/class-fixinsecurecontent.php]: As I commented on the other conversation, the functionality to bulk scan/fix insecure content from the admin is actually a replication of the WP CLI functionality. The WP CLI functionality can properly scan/fix post types that are not shown in REST, so in my opinion, the admin functionality should behave exactly the same. According to my tests, neither the WP CLI nor the built admin functionality has problems scanning/fixing posts from a post type that is not shown in REST.

File: includes/classes/class-fixinsecurecontent.php
Comment (kmgalanakis) [reply includes/classes/class-fixinsecurecontent.php]: This is part of the code from the implementation of the WP CLI command but I can confirm that your point is right, it checks if the URL resolves to an existing resource without downloading it just by checking the returned HTTP header.

Good point, thank you.

File: src/js/classic-editor.js
Comment (kmgalanakis) [reply src/js/classic-editor.js:5]: I think you mean `import jQuery from 'jquery';`. Initially, this didn't play nicely with ESLint but I ended up copying the way it's done in `Distributor` and now it's fine.

Thank you.

File: src/js/gutenberg.js
Comment (kmgalanakis) [reply src/js/gutenberg.js:10]: Good catch, thank you.

File: src/js/utils/blur-insecure.js
Comment (kmgalanakis) [reply src/js/utils/blur-insecure.js:1]: Good catch, thank you.

File: src/js/utils/check-content.js
Comment (kmgalanakis) [reply src/js/utils/check-content.js:5]: Good catch, thank you.

File: src/js/utils/find-elements.js
Comment (kmgalanakis) [reply src/js/utils/find-elements.js:3]: Good catch, thank you.

File: src/js/utils/replace.js
Comment (kmgalanakis) [reply src/js/utils/replace.js:1]: Good catch, thank you.

File: includes/classes/class-fixinsecurecontent.php
Comment (peterwilsoncc) [reply includes/classes/class-fixinsecurecontent.php]: > The WP CLI functionality can properly scan/fix post types that are not shown in REST, so in my opinion, the admin functionality should behave exactly the same.

The difference between the two is that the WP-CLI is basically the equivalent of running as the `root` user on a server. Once someone has access to the server they can do anything, such as running `wp db clean --yes` so permissions don't apply.

For users with access to the dashboard only, we need to check that they have full permission to make any changes to a post. If a post type isn't editable via the rest API then we need to assume the developer has reasons for preventing this and respect their decision

Comment (kmgalanakis) [reply includes/classes/class-fixinsecurecontent.php]: What you say makes perfect sense @peterwilsoncc. I've modified the method that counts posts with and the method that fixes insecure content so that posts from post types that are not shown in RFEST are excluded from the process.

Please let me know if this is what you meant. Thank you for your feedback and patience.

File: includes/classes/class-fixinsecurecontent.php
Comment (peterwilsoncc) [inline includes/classes/class-fixinsecurecontent.php]:
```suggestion
```

I think the other two will be fine, the user may have permission to edit some private post types

File: includes/classes/class-fixinsecurecontent.php
Comment (peterwilsoncc) [inline includes/classes/class-fixinsecurecontent.php]: you've used `all` elsewhere and `any` here.

**Review by peterwilsoncc — COMMENTED** [review]
Looks good with a minor nitpcik.

File: includes/classes/class-fixinsecurecontent.php
Comment (kmgalanakis) [reply includes/classes/class-fixinsecurecontent.php]: Done.

File: includes/classes/class-fixinsecurecontent.php
Comment (kmgalanakis) [reply includes/classes/class-fixinsecurecontent.php]: Good point, fixed.

**Review by peterwilsoncc — DISMISSED** [review]
LGTM, thank you!

**Review by peterwilsoncc — APPROVED** [review]
Resolving merge conflicts dismissed my review, reapproving.

**Pull request merged**
